src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1134,26 +1170,56 @@ private void ImportCompareOperation(ILOpcode opcode)
             }
 
             LLVMValueRef result;
-            switch (opcode)
+            var typeSaneOp1 = op1.LLVMValue;
+            var typeSaneOp2 = op2.LLVMValue;
+            if (op1.Type != op2.Type || op1.Type == null)
             {
-                case ILOpcode.ceq:
-                    result = LLVM.BuildICmp(_builder, LLVMIntPredicate.LLVMIntEQ, op2.LLVMValue, op1.LLVMValue, "ceq");
-                    break;
-                case ILOpcode.cgt:
-                    result = LLVM.BuildICmp(_builder, LLVMIntPredicate.LLVMIntSGT, op2.LLVMValue, op1.LLVMValue, "cgt");
-                    break;
-                case ILOpcode.clt:
-                    result = LLVM.BuildICmp(_builder, LLVMIntPredicate.LLVMIntSLT, op2.LLVMValue, op1.LLVMValue, "clt");
-                    break;
-                case ILOpcode.cgt_un:
-                    result = LLVM.BuildICmp(_builder, LLVMIntPredicate.LLVMIntUGT, op2.LLVMValue, op1.LLVMValue, "cgt_un");
-                    break;
-                case ILOpcode.clt_un:
-                    result = LLVM.BuildICmp(_builder, LLVMIntPredicate.LLVMIntULT, op2.LLVMValue, op1.LLVMValue, "clt_un");
-                    break;
-                default:
-                    throw new NotSupportedException(); // unreachable
+                if (op1.Type != null && op2.Type != null)
+                {
+                    if (op1.Type.IsPrimitive && op2.Type.IsPrimitive)
+                    {
+                        if (op1.Type.GetElementSize().AsInt > op2.Type.GetElementSize().AsInt)
```

Comment:
We probably have to account for sign extension here. This PR is already pretty big, so I'd be ok with a comment and filing an issue if you prefer.

Reply:
I've added a todo comment above calling out the lack of sign extension